scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/TreeOrderedBuf.scala

```diff
@@ -0,0 +1,309 @@
+      val tmpArray = freshT("tmpArray")
+      val len = freshT("len")
+      val oldPos = freshT("oldPos")
+      val noLenCalc = q"""
+
+      val $tmpArray = {
+        val baos = new _root_.java.io.ByteArrayOutputStream
+        innerPutNoLen(baos, $element)
+        baos.toByteArray
+      }
+
+      val $len = $tmpArray.size
+      $outerbaos.writeSize($len)
+      $outerbaos.writeBytes($tmpArray)
+      """
+
+      def withLenCalc(lenC: Tree) = q"""
+        val $len = $lenC
+        $outerbaos.writeSize($len)
+        innerPutNoLen($outerbaos, $element)
+      """
+
+      t.length(q"$element") match {
+        case _: NoLengthCalculationAvailable[_] => noLenCalc
+        case _: ConstantLengthCalculation[_] => q"""
+        innerPutNoLen($outerbaos, $element)
+        """
+        case f: FastLengthCalculation[_] =>
+          withLenCalc(f.asInstanceOf[FastLengthCalculation[c.type]].t)
+        case m: MaybeLengthCalculation[_] =>
+          val tmpLenRes = freshT("tmpLenRes")
+          q"""
+            def noLenCalc = {
+              $noLenCalc
+            }
+            def withLenCalc(cnt: Int) = {
+              ${withLenCalc(q"cnt")}
+            }
+            val $tmpLenRes: _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.MaybeLength = payloadLength($element)
+            $tmpLenRes match {
+              case _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.NoLengthCalculation => noLenCalc
+              case _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.ConstLen(const) => withLenCalc(const)
+              case _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.DynamicLen(s) => withLenCalc(s)
+            }
+        """
+      }
+    }
+
+    def readLength(inputStream: TermName) = {
+      t.length(q"e") match {
+        case const: ConstantLengthCalculation[_] => q"${const.toInt}"
+        case _ => q"$inputStream.readSize"
+      }
+    }
+
+    def discardLength(inputStream: TermName) = {
+      t.length(q"e") match {
+        case const: ConstantLengthCalculation[_] => q"()"
+        case _ => q"$inputStream.readSize"
+      }
+    }
+
+    val lazyVariables = t.lazyOuterVariables.map {
+      case (n, t) =>
+        val termName = newTermName(n)
+        q"""lazy val $termName = $t"""
+    }
+
+    val element = freshT("element")
+
+    val inputStreamA = freshT("inputStreamA")
+    val inputStreamB = freshT("inputStreamB")
+
+    val lenA = freshT("lenA")
+    val lenB = freshT("lenB")
+
+    t.ctx.Expr[OrderedSerialization[T]](q"""
+      new _root_.com.twitter.scalding.serialization.OrderedSerialization[$T] with _root_.com.twitter.bijection.macros.MacroGenerated  {
+        import com.twitter.scalding.serialization.JavaStreamEnrichments._
+        ..$lazyVariables
+
+        final def innerBinaryCompare($inputStreamA: _root_.java.io.InputStream, $inputStreamB: _root_.java.io.InputStream): Int = {
```

Comment:
if you make this a final val that is a function2, you won't have to box it again on every compare (the method to function takes a boxing).